#<b><span style='color:#3A6D8C'>CLUSTERING</span></b>

Clustering adalah metode dalam analisis data yang digunakan untuk mengelompokkan data ke dalam grup atau kluster berdasarkan kesamaan di antara data tersebut. Tujuannya adalah untuk mengidentifikasi pola atau struktur dalam dataset tanpa adanya label atau informasi sebelumnya mengenai kelompok mana yang harus dihasilkan.

#<b>A. <span style='color:#3A6D8C'>Data</span></b>

### <b>A.1. <span style='color:#3A6D8C'>Ekstrak Data</span></b>

In [ ]:
import warnings
warnings.filterwarnings("ignore")

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# File path - Modified to access raw data
url = "https://raw.githubusercontent.com/MTOZZ/data_aspek_ikp/refs/heads/main/aspek_ikp_2022.csv"
  # Updated URL for raw data access

# Membaca file CSV dengan opsi tambahan untuk memastikan semua data terbaca
data_aspek_indonesia = pd.read_csv(
    url,
    delimiter=',',          # Sesuaikan delimiter (default: koma)
    encoding='utf-8',       # Encoding umum untuk file CSV
)

data_aspek_indonesia["tahun"] = 2022

# Menampilkan 10 baris pertama untuk verifikasi
display(data_aspek_indonesia.head())

provinsi  ketersediaan  keterjangkauan  pemanfaatan  tahun
0            Aceh         80.17           69.34        66.25   2022
1  Sumatera Utara         80.19           75.02        64.44   2022
2  Sumatera Barat         87.53           84.23        71.84   2022
3            Riau         38.78           86.50        69.81   2022
4           Jambi         58.48           82.00        67.04   2022

### <b>A.2. <span style='color:#3A6D8C'>Informasi Umum Data</span></b>

In [ ]:
data_aspek_indonesia.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   provinsi        34 non-null     object 
 1   ketersediaan    34 non-null     float64
 2   keterjangkauan  34 non-null     float64
 3   pemanfaatan     34 non-null     float64
 4   tahun           34 non-null     int64  
dtypes: float64(3), int64(1), object(1)
memory usage: 1.5+ KB


#<b>B. <span style='color:#3A6D8C'>Exploratory Data Analysis</span></b>

### <b>B.1. <span style='color:#3A6D8C'>Filter Data dan Split Data</span></b>


In [ ]:
# Inisialisasi kolom
kolom = [
    'provinsi',
    'tahun',
    'ketersediaan',
    'keterjangkauan',
    'pemanfaatan'

]

# Ambil kolom yang ditentukan saja
data_cluster_indo = data_aspek_indonesia[kolom]

# Tampilkan hasilnya
display(data_cluster_indo)

provinsi  tahun  ketersediaan  keterjangkauan  pemanfaatan
0                   Aceh   2022         80.17           69.34        66.25
1         Sumatera Utara   2022         80.19           75.02        64.44
2         Sumatera Barat   2022         87.53           84.23        71.84
3                   Riau   2022         38.78           86.50        69.81
4                  Jambi   2022         58.48           82.00        67.04
5       Sumatera Selatan   2022         81.57           74.11        61.29
6               Bengkulu   2022         69.92           75.18        61.53
7                Lampung   2022         94.94           79.62        67.72
8   Kep. Bangka Belitung   2022         36.38           92.47        77.54
9         Kepulauan Riau   2022          5.83           88.90        72.53
10           DKI Jakarta   2022          0.00           83.22        81.12
11            Jawa Barat   2022         84.59           79.11        73.71
12           Jawa Tengah   2022         88.88           81.47        80.69
13         DI Yogyakarta   2022         84.20           79.66        79.97
14            Jawa Timur   2022         90.64           79.42        74.27
15                Banten   2022         82.81           83.99        62.79
16                  Bali   2022         79.08           94.00        82.08
17   Nusa Tenggara Barat   2022         92.85           76.24        67.81
18   Nusa Tenggara Timur   2022         86.25           61.62        60.55
19      Kalimantan Barat   2022         76.21           86.33        56.29
20     Kalimantan Tengah   2022         49.87           91.07        67.55
21    Kalimantan Selatan   2022         88.69           90.68        69.01
22      Kalimantan Timur   2022         52.33           90.27        79.43
23      Kalimantan Utara   2022         43.50           89.99        72.21
24        Sulawesi Utara   2022         70.37           81.42        71.42
25       Sulawesi Tengah   2022         77.98           80.39        70.88
26      Sulawesi Selatan   2022         94.71           83.52        71.22
27     Sulawesi Tenggara   2022         71.80           78.24        74.38
28             Gorontalo   2022         94.25           79.76        71.91
29        Sulawesi Barat   2022         76.11           83.45        65.42
30                Maluku   2022         45.87           68.57        62.27
31          Maluku Utara   2022         14.50           88.12        60.61
32           Papua Barat   2022         11.81           62.11        56.17
33                 Papua   2022         17.68           49.89        42.34

In [ ]:
# Lakukan filter
data_cluster_aspek_2022 = data_cluster_indo[data_cluster_indo['tahun'] == 2022]

# Tampilkan hasilnya
data_cluster_aspek_2022.head()

provinsi  tahun  ketersediaan  keterjangkauan  pemanfaatan
0            Aceh   2022         80.17           69.34        66.25
1  Sumatera Utara   2022         80.19           75.02        64.44
2  Sumatera Barat   2022         87.53           84.23        71.84
3            Riau   2022         38.78           86.50        69.81
4           Jambi   2022         58.48           82.00        67.04

Melakukan filter data aspek ketahanan pangan berdasarkan tahun 2022

### <b>B.2. <span style='color:#3A6D8C'>Identifikasi Missing Value</span></b>

In [ ]:
# Buat dataframe kosong
data_kosong = pd.DataFrame(index = data_cluster_aspek_2022.columns.to_list())

# Lakukan perhitungan
for provinsi in data_cluster_aspek_2022['provinsi'].unique():
    condition = (data_cluster_aspek_2022['provinsi'] == provinsi)
    list_banyak_null = list()
    for col in data_cluster_aspek_2022.columns.to_list():
        banyak_nul = data_cluster_aspek_2022.loc[condition][col].isna().sum()
        list_banyak_null.append(banyak_nul)
    data_kosong[provinsi] = list_banyak_null

# Buat highlight merah untuk missing value > 0
styled_df = data_kosong.T.style.map(lambda val : 'background-color: red' if val > 0 else '')

display(styled_df)

Identifikasi missing value, data aspek ketahan pangan tidak terdapat missing value

### <b>B.3. <span style='color:#3A6D8C'>Identifikasi Duplikasi Data</span></b>

In [ ]:
data_cluster_2022 = data_cluster_aspek_2022.drop_duplicates()

print(f'Jumlah duplikasi data = {data_cluster_2022.duplicated().sum()}')

Jumlah duplikasi data = 0


Tidak ada duplikasi dalam data aspek ketahanan pangan tahun 2022

### <b>B.4. <span style='color:#3A6D8C'>Identifikasi Outlier</span></b>

In [ ]:
# Histogram Sales
import plotly.express as px

col_name = ['ketersediaan', 'keterjangkauan', 'pemanfaatan']
color_name = ['#FF8F00']

for column, color in zip(col_name, color_name):
    fig = px.box(
        data_cluster_2022,
        x = col_name,
        color_discrete_sequence  = [color],
    )

    fig.update_traces(
          marker_line_width = 1,
          marker_line_color = 'white'
    )

    fig.update_layout(
        width=1000,
        height=600,
        plot_bgcolor='rgba(0, 0, 0, 0)',
        title=dict(
            text=f"<b>Box Plot: <span style='color:{color}; font-size: 15px;'>Outliers Analysis</span>",
            font=dict(
                size=20,
                color='#757882'
            ),
            y=0.92,
            x=0.5
        ),
        yaxis=dict(
            title='',
            showgrid=True,
            zeroline=True,
        ),
        margin = dict(
            t = 80,
            b = 10,
            r = 20
        )
    )

    fig.show(renderer = 'colab')

Identifikasi *outlier* dengan boxplot menunjukkan bahwa terdapat *outlier *pada data pemanfaatan dan keterjangkauan dengan nilai *outlier* pemanfaatan **42.34** dan nilai *outlier* keterjangkauan **49.89**

#### <b>B.4.1. <span style='color:#3A6D8C'>Penanganan Outlier</span></b>

In [ ]:
def replace_outliers_iqr(df):

    # Menghitung Q1 dan Q3
    Q1 = df.quantile(0.25)
    Q3 = df.quantile(0.75)
    IQR = Q3 - Q1

    # Menentukan batas bawah dan atas
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR

    # Mengganti outlier
    df = np.select(
        [df > upper_bound, df < lower_bound],
        [upper_bound, lower_bound],
        default = df
    )

    return df

In [ ]:
# Daftar kolom yang ingin ditangani outlier-nya
columns_to_handle = ['pemanfaatan','keterjangkauan']  # Tambahkan kolom lain jika diperlukan, misalnya ['ikp', 'sales']

# Terapkan fungsi replace_outliers_iqr hanya pada kolom tertentu
for col in columns_to_handle:
    data_cluster_2022[col] = replace_outliers_iqr(data_cluster_2022[col])

# Visualisasi box plot setelah menangani outliers
import plotly.express as px

for column in columns_to_handle:
    fig = px.box(
        data_cluster_2022,
        x=column,
        color_discrete_sequence=['#FF8F00'],
    )

    fig.update_traces(
        marker_line_width=1,
        marker_line_color='white'
    )

    fig.update_layout(
        width=1000,
        height=600,
        plot_bgcolor='rgba(0, 0, 0, 0)',
        title=dict(
            text=f"<b>Box Plot<br><span style='color:#FF8F00; font-size: 15px;'>{column}</b>",
            font=dict(
                size=28,
                color='#757882'
            ),
            y=0.92,
            x=0.5
        ),
        yaxis=dict(
            title='',
            showgrid=False,
            showline=False,
            showticklabels=False,
            zeroline=False,
        ),
        margin=dict(
            t=80,
            b=10,
            r=20
        )
    )

    fig.show(renderer='colab')


Sebelumnya, terdapat *outlier* pada data pemanfaatan dan keterjangkauan. Oleh karena itu, perlu adanya penanganan pada *outlier* tersebut agar hasil analisisnya lebih akurat. Berdasarkan boxplot tersebut, dapat dilihat bahwa sudah tidak ada *outlier*, sehingga penanganan *outlier* berhasil dilakukan.

### <b>B.5. <span style='color:#3A6D8C'>Scaling Data</span></b>

In [ ]:
# Import library yang dibutuhkan
from sklearn.preprocessing import MinMaxScaler

# Normalisasi data menggunakan StandardScaler
scaler = MinMaxScaler()
features = ['ketersediaan', 'keterjangkauan', 'pemanfaatan']
normalized_data = scaler.fit_transform(data_cluster_2022[features])

# Membuat DataFrame dari data yang telah dinormalisasi
normalized_df = pd.DataFrame(normalized_data, columns=[f"{col} normalize" for col in features])

# Menggabungkan kembali dengan kolom provinsi
normalized_df['provinsi'] = data_cluster_2022['provinsi']

# Menampilkan data yang telah dinormalisasi untuk verifikasi
normalized_df.head()

ketersediaan normalize  keterjangkauan normalize  pemanfaatan normalize  \
0                0.844428                  0.225411               0.537084   
1                0.844639                  0.403824               0.484154   
2                0.921951                  0.693117               0.700552   
3                0.408469                  0.764419               0.641189   
4                0.615968                  0.623071               0.560186   

         provinsi  
0            Aceh  
1  Sumatera Utara  
2  Sumatera Barat  
3            Riau  
4           Jambi

Hasil normalisasi menggunakan MinMax Scaler pada data di atas menunjukkan bahwa setiap nilai pada kolom ketersediaan, keterjangkauan, dan pemanfaatan telah diubah ke rentang 0 hingga 1 berdasarkan nilai minimum dan maksimum masing-masing kolom.

Normalisasi data ini membantu memudahkan analisis perbandingan antarprovinsi dalam rentang yang sama, sehingga dapat digunakan untuk pengelompokan atau rekomendasi kebijakan berdasarkan indikator ketahanan pangan.

In [ ]:
# Proses masukan data
for col in normalized_df.columns:
    data_cluster_2022[col] = normalized_df[col]

# Tampilkan hasilnya
display(data_cluster_2022)

provinsi  tahun  ketersediaan  keterjangkauan  pemanfaatan  \
0                   Aceh   2022         80.17        69.34000     66.25000   
1         Sumatera Utara   2022         80.19        75.02000     64.44000   
2         Sumatera Barat   2022         87.53        84.23000     71.84000   
3                   Riau   2022         38.78        86.50000     69.81000   
4                  Jambi   2022         58.48        82.00000     67.04000   
5       Sumatera Selatan   2022         81.57        74.11000     61.29000   
6               Bengkulu   2022         69.92        75.18000     61.53000   
7                Lampung   2022         94.94        79.62000     67.72000   
8   Kep. Bangka Belitung   2022         36.38        92.47000     77.54000   
9         Kepulauan Riau   2022          5.83        88.90000     72.53000   
10           DKI Jakarta   2022          0.00        83.22000     81.12000   
11            Jawa Barat   2022         84.59        79.11000     73.71000   
12           Jawa Tengah   2022         88.88        81.47000     80.69000   
13         DI Yogyakarta   2022         84.20        79.66000     79.97000   
14            Jawa Timur   2022         90.64        79.42000     74.27000   
15                Banten   2022         82.81        83.99000     62.79000   
16                  Bali   2022         79.08        94.00000     82.08000   
17   Nusa Tenggara Barat   2022         92.85        76.24000     67.81000   
18   Nusa Tenggara Timur   2022         86.25        62.16375     60.55000   
19      Kalimantan Barat   2022         76.21        86.33000     56.29000   
20     Kalimantan Tengah   2022         49.87        91.07000     67.55000   
21    Kalimantan Selatan   2022         88.69        90.68000     69.01000   
22      Kalimantan Timur   2022         52.33        90.27000     79.43000   
23      Kalimantan Utara   2022         43.50        89.99000     72.21000   
24        Sulawesi Utara   2022         70.37        81.42000     71.42000   
25       Sulawesi Tengah   2022         77.98        80.39000     70.88000   
26      Sulawesi Selatan   2022         94.71        83.52000     71.22000   
27     Sulawesi Tenggara   2022         71.80        78.24000     74.38000   
28             Gorontalo   2022         94.25        79.76000     71.91000   
29        Sulawesi Barat   2022         76.11        83.45000     65.42000   
30                Maluku   2022         45.87        68.57000     62.27000   
31          Maluku Utara   2022         14.50        88.12000     60.61000   
32           Papua Barat   2022         11.81        62.16375     56.17000   
33                 Papua   2022         17.68        62.16375     47.88375   

    ketersediaan normalize  keterjangkauan normalize  pemanfaatan normalize  
0                 0.844428                  0.225411               0.537084  
1                 0.844639                  0.403824               0.484154  
2                 0.921951                  0.693117               0.700552  
3                 0.408469                  0.764419               0.641189  
4                 0.615968                  0.623071               0.560186  
5                 0.859174                  0.375240               0.392039  
6                 0.736465                  0.408850               0.399057  
7                 1.000000                  0.548314               0.580071  
8                 0.383189                  0.951942               0.867237  
9                 0.061407                  0.839805               0.720730  
10                0.000000                  0.661392               0.971927  
11                0.890984                  0.532294               0.755236  
12                0.936170                  0.606423               0.959352  
13                0.886876                  0.549570               0.938297  
14                0.954708                  0.542031               0.771612  
15                0.872235           

#<b>C. <span style='color:#3A6D8C'>Modelling dengan K-Means</span></b>
K-Means adalah algoritma unsupervised machine learning yang digunakan untuk clustering atau pengelompokan data. Algoritma ini bekerja dengan membagi data ke dalam sejumlah kelompok (cluster) berdasarkan kemiripan atau kedekatan fitur (atribut) dalam ruang multidimensi. Tujuannya adalah meminimalkan variasi dalam cluster dan memaksimalkan perbedaan antar cluster.


### <b>C.1. <span style='color:#3A6D8C'>Identifikasi Silhouette Score</span></b>

In [ ]:
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score

# Definisikan kolom yang akan di clustering
X = data_cluster_2022[[
    'ketersediaan normalize',
    'keterjangkauan normalize',
    'pemanfaatan normalize'
]]

k_range = range(2, 16)
silhouette = list()

# Melakukan K-Means dan menghitung Silhouette Score
for k in k_range:
    kmeans = KMeans(n_clusters=k, random_state=0)
    cluster_labels = kmeans.fit_predict(X)
    score = silhouette_score(X, cluster_labels)
    silhouette.append(score)

In [ ]:
import plotly.express as px

# Plot grafik
fig = px.line(
    x = k_range,
    y = silhouette,
    color_discrete_sequence = ['#FFAD60']
)

fig.update_layout(
    width = 1200,
    height = 500,
    showlegend = False,
    plot_bgcolor = 'rgba(0, 0, 0, 0)',
    title = dict(
        text = "<b>Nilai <span style='color:#FFAD60'>Silhouette Score</span>\
        </b><br><sup><sup>dengan Algoritma K-Means</sup></sup>",
        font = dict(
            size = 25,
            color = '#757882'
        ),
    ),
    xaxis = dict(
        showgrid = False,
        showline = True,
        linecolor = 'black',
        showticklabels = True,
        zeroline = True,
    ),
    yaxis = dict(
        showgrid = False,
        showline = True,
        linecolor='black',
        showticklabels = True,
        zeroline = True,
    ),
    margin = dict(
        l = 50
    )
)

fig.show(renderer = 'colab')

Dari hasil Silhouette Score didapatkan jumlah cluster terbaik yaitu (K) = 4 dengan nilai sebesar **0.3791**

### <b>C.2. <span style='color:#3A6D8C'>Penentuan Jumlah Cluster</span></b>

In [ ]:
kmeans = KMeans(n_clusters = 4, random_state = 0)
cluster_labels = kmeans.fit_predict(X)

data_cluster_2022['cluster_kmeans'] = cluster_labels + 1
display(data_cluster_2022)

provinsi  tahun  ketersediaan  keterjangkauan  pemanfaatan  \
0                   Aceh   2022         80.17        69.34000     66.25000   
1         Sumatera Utara   2022         80.19        75.02000     64.44000   
2         Sumatera Barat   2022         87.53        84.23000     71.84000   
3                   Riau   2022         38.78        86.50000     69.81000   
4                  Jambi   2022         58.48        82.00000     67.04000   
5       Sumatera Selatan   2022         81.57        74.11000     61.29000   
6               Bengkulu   2022         69.92        75.18000     61.53000   
7                Lampung   2022         94.94        79.62000     67.72000   
8   Kep. Bangka Belitung   2022         36.38        92.47000     77.54000   
9         Kepulauan Riau   2022          5.83        88.90000     72.53000   
10           DKI Jakarta   2022          0.00        83.22000     81.12000   
11            Jawa Barat   2022         84.59        79.11000     73.71000   
12           Jawa Tengah   2022         88.88        81.47000     80.69000   
13         DI Yogyakarta   2022         84.20        79.66000     79.97000   
14            Jawa Timur   2022         90.64        79.42000     74.27000   
15                Banten   2022         82.81        83.99000     62.79000   
16                  Bali   2022         79.08        94.00000     82.08000   
17   Nusa Tenggara Barat   2022         92.85        76.24000     67.81000   
18   Nusa Tenggara Timur   2022         86.25        62.16375     60.55000   
19      Kalimantan Barat   2022         76.21        86.33000     56.29000   
20     Kalimantan Tengah   2022         49.87        91.07000     67.55000   
21    Kalimantan Selatan   2022         88.69        90.68000     69.01000   
22      Kalimantan Timur   2022         52.33        90.27000     79.43000   
23      Kalimantan Utara   2022         43.50        89.99000     72.21000   
24        Sulawesi Utara   2022         70.37        81.42000     71.42000   
25       Sulawesi Tengah   2022         77.98        80.39000     70.88000   
26      Sulawesi Selatan   2022         94.71        83.52000     71.22000   
27     Sulawesi Tenggara   2022         71.80        78.24000     74.38000   
28             Gorontalo   2022         94.25        79.76000     71.91000   
29        Sulawesi Barat   2022         76.11        83.45000     65.42000   
30                Maluku   2022         45.87        68.57000     62.27000   
31          Maluku Utara   2022         14.50        88.12000     60.61000   
32           Papua Barat   2022         11.81        62.16375     56.17000   
33                 Papua   2022         17.68        62.16375     47.88375   

    ketersediaan normalize  keterjangkauan normalize  pemanfaatan normalize  \
0                 0.844428                  0.225411               0.537084   
1                 0.844639                  0.403824               0.484154   
2                 0.921951                  0.693117               0.700552   
3                 0.408469                  0.764419               0.641189   
4                 0.615968                  0.623071               0.560186   
5                 0.859174                  0.375240               0.392039   
6                 0.736465                  0.408850               0.399057   
7                 1.000000                  0.548314               0.580071   
8                 0.383189                  0.951942               0.867237   
9                 0.061407                  0.839805               0.720730   
10                0.000000                  0.661392               0.971927   
11                0.890984                  0.532294               0.755236   
12                0.936170                  0.606423               0.959352   
13                0.886876                  0.549570               0.938297   
14                0.954708                  0.542031               0.771612   
15                0.8

Berdasarkan output tersebut, dapat dilihat bahwa setiap provinsi berhasil
dikelompokkan ke dalam 4 cluster yang berbeda, yaitu cluster 1, 2, 3 dan 4. Hal ini menunjukkan bahwa algoritma K-Means berhasil mengelompokkan provinsi-provinsi berdasarkan kesamaan karakteristiknya.

### <b>C.3. <span style='color:#3A6D8C'>Silhouette Plot</span></b>

In [ ]:
import numpy as np
import pandas as pd
import plotly.graph_objects as go
from sklearn.metrics import silhouette_samples, silhouette_score
from sklearn.cluster import KMeans

# Hitung silhouette score dan nilai untuk setiap titik
silhouette_avg = silhouette_score(X, cluster_labels)
silhouette_values = silhouette_samples(X, cluster_labels)

# Siapkan data untuk visualisasi
n_clusters = len(np.unique(cluster_labels))
silhouette_data = pd.DataFrame({
    'Cluster': cluster_labels + 1,
    'Silhouette Coefficient': silhouette_values
})

# Urutkan data berdasarkan klaster dan nilai silhouette
silhouette_data = silhouette_data.sort_values(by=['Cluster', 'Silhouette Coefficient'], ascending=[True, True])

# Tambahkan kolom posisi untuk membantu visualisasi
silhouette_data['Posisi'] = np.arange(len(silhouette_data))

# Membuat plot menggunakan Plotly
fig = go.Figure()

# Loop untuk setiap klaster
for cluster_id in range(1, n_clusters + 1):
    cluster_data = silhouette_data[silhouette_data['Cluster'] == cluster_id]
    fig.add_trace(go.Bar(
        x=cluster_data['Silhouette Coefficient'],
        y=cluster_data['Posisi'],
        orientation='h',
        marker=dict(color=px.colors.qualitative.Plotly[cluster_id % len(px.colors.qualitative.Plotly)]),
        name=f'Cluster {cluster_id}'
    ))

# Garis rata-rata silhouette score
fig.add_vline(
    x=silhouette_avg,
    line=dict(color='red', dash='dash'),
    annotation_text="Rata-rata Silhouette",
    annotation_position="top right"
)

# Layout
fig.update_layout(
    width=1200,
    height=600,
    title=dict(
        text="<b>Silhouette Plot</b><br><sup><sup>dengan Algoritma K-Means</sup></sup>",
        font=dict(size=25, color='#757882'),
    ),
    xaxis=dict(
        title="Nilai Silhouette Coefficient",
        linecolor="black",
        showgrid=False,
        zeroline=True
    ),
    yaxis=dict(
        title="Cluster",
        showticklabels=False,
        linecolor="black",
        showgrid=False
    ),
    plot_bgcolor='rgba(0, 0, 0, 0)',
    showlegend=False,
    margin=dict(l=50)
)

fig.show(renderer='colab')


Dari hasil Silhouette Plot, rata-rata Silhouette Score bernilai **0.37** nilai ini menunjukkan pengelompokkan cukup baik tetapi tidak optimal, karena mendekati batas bawah moderat.

### <b>C.4. <span style='color:#3A6D8C'>Plot Hasil Cluster</span></b>

In [ ]:
import plotly.graph_objects as go

# Membuat plot 3D
fig = go.Figure(
    data = [
        go.Scatter3d(
            x = data_cluster_2022['ketersediaan'],
            y = data_cluster_2022['keterjangkauan'],
            z = data_cluster_2022['pemanfaatan'],
            mode = 'markers',
            marker = dict(
                size = 5,
                color = data_cluster_2022['cluster_kmeans'],
                opacity = 0.5,
                colorscale='Viridis',
                colorbar = dict(title='Color Legend')
            ),
            hovertext = data_cluster_2022['provinsi'],
            hovertemplate='<b>%{hovertext}</b><br>ketersediaan: %{x}<br>Keterjangkauan: %{y}<br>Pemanfaatan: %{z}<br>Cluster: %{marker.color}'
        )
    ]
)


fig.update_layout(
    width = 1200,
    height = 800,
    plot_bgcolor = 'rgba(0, 0, 0, 0)',
    title = dict(
        text = '',
        font = dict(
            size = 20
        )
    ),
    scene = dict(
        xaxis_title = 'Ketersediaan',
        yaxis_title = 'Keterjangkauaan',
        zaxis_title = 'Pemanfaatan',
        xaxis=dict(showgrid=False),
        yaxis=dict(showgrid=False),
        zaxis=dict(showgrid=False)
    )
)

# Tampilkan plot
fig.show()

Proses clustering ini menghasilkan 4 kelompok (cluster) berbeda berdasarkan karakteristik kesamaan dalam ketiga aspek tersebut. Setelah dilakukan clustering, setiap provinsi diberi label sesuai dengan kelompoknya. Berdasarkan hasil clustering, berikut adalah pembagian provinsi dalam masing-masing cluster:

Cluster 1 (Ungu): Daerah dengan ketersediaan dan keterjangkauan yang tinggi serta pemanfaatan yang stabil, meskipun terdapat variasi sedang pada ketersediaan.

Cluster 2 (Biru): Daerah dengan keterjangkauan yang sangat tinggi dan konsisten, tingkat pemanfaatan yang baik, namun tingkat ketersediaan masih rendah.

Cluster 3 (Hijau): Daerah dengan ketersediaan, keterjangkauan, dan pemanfaatan menunjukkan tingkat yang tinggi pada ketiganya, dengan variasi kecil hingga sedang.

Cluster 4 (Kuning) : Daerah dengan tingkat ketersediaan yang sangat rendah, keterjangkauan yang moderat tanpa variasi, dan pemanfaatan yang rendah hingga sedang dengan variasi kecil.

Hasil ini dapat digunakan untuk mengidentifikasi daerah yang membutuhkan perhatian khusus seperti daerah di Cluster 4.
Selain itu, cluster dengan nilai tinggi seperti Cluster 3 dapat dijadikan contoh atau perbandingan untuk daerah lainnya.

### <b>C.5. <span style='color:#3A6D8C'>Summary</span></b>

In [ ]:
# Menghitung mean, median, dan simpangan baku untuk setiap cluster
cluster_statistics = data_cluster_2022.groupby('cluster_kmeans').agg({
    'ketersediaan': ['mean', 'median', 'std'],
    'keterjangkauan': ['mean', 'median', 'std'],
    'pemanfaatan': ['mean', 'median', 'std']
}).reset_index()

# Menampilkan hasil
display(cluster_statistics)

cluster_kmeans ketersediaan                    keterjangkauan            \
                         mean  median        std           mean    median   
0              1    73.995000  80.180  14.776302      70.730625  71.72500   
1              2    30.148750  37.580  20.416207      88.817500  89.44500   
2              3    83.006667  84.395   9.921597      82.418333  81.44500   
3              4    14.745000  14.745   4.150717      62.163750  62.16375   

            pemanfaatan                       
        std        mean     median       std  
0  5.091352   62.721667  61.900000  2.181361  
1  2.905540   72.600000  72.370000  6.770172  
2  4.405731   71.025000  71.320000  6.342421  
3  0.000000   52.026875  52.026875  5.859264

**Cluster 1 (Ungu)**

Ketersediaan rata-rata 73,99 (median 80,18), keterjangkauan 70,73 (median 71,73), dan pemanfaatan 62,72 (median 61,90). Menunjukkan wilayah dengan ketersediaan tinggi, keterjangkauan baik, dan pemanfaatan stabil, mencerminkan infrastruktur atau akses yang baik.


**Cluster 2 (Biru)**

Ketersediaan rata-rata 30,15 (median 37,58), keterjangkauan 88,82 (median 89,45), dan pemanfaatan 72,60 (median 72,37). Menunujukkan wilayah dengan keterjangkauan sangat tinggi, namun ketersediaan rendah membatasi, meskipun pemanfaatan cukup baik berkat optimasi sumber daya.

**Cluster 3 (Hijau)**

Ketersediaan rata-rata 83,01 (median 84,39), keterjangkauan 82,42 (median 81,45), dan pemanfaatan 71,03 (median 71,32). Menunujukkan wilayah dengan keseimbangan optimal antara ketersediaan, keterjangkauan, dan pemanfaatan sumber daya.

**Cluster 4 (Kuning)**

Ketersediaan rata-rata 14,75, keterjangkauan 62,16, dan pemanfaatan 52,03 (dengan median yang sama)  Hal ini mengindikasikan area ini memiliki akses yang terbatas, meskipun keterjangkauannya stabil.

In [ ]:
!pip install scikit-learn-extra

#<b>D. <span style='color:#3A6D8C'>Modelling dengan K-Medoids</span></b>

Proses modelling dilakukan menggunakan K-Medoids. K-Medoids ialah algoritma clustering yang hampir sama dengan algoritma K-Means. Perbedaannya adalah K-Medoids mengunakan objek sebagai pusat cluster untuk setiap cluster, sedangkan K-Means menggunakan nilai rata-rata sebagai pusat cluster untuk setiap cluster.

### <b>D.1. <span style='color:#3A6D8C'>Proses Scalling</span></b>

In [ ]:
data_cluster_kmedoids = data_cluster_2022.copy()
display(data_cluster_kmedoids)

provinsi  tahun  ketersediaan  keterjangkauan  pemanfaatan  \
0                   Aceh   2022         80.17        69.34000     66.25000   
1         Sumatera Utara   2022         80.19        75.02000     64.44000   
2         Sumatera Barat   2022         87.53        84.23000     71.84000   
3                   Riau   2022         38.78        86.50000     69.81000   
4                  Jambi   2022         58.48        82.00000     67.04000   
5       Sumatera Selatan   2022         81.57        74.11000     61.29000   
6               Bengkulu   2022         69.92        75.18000     61.53000   
7                Lampung   2022         94.94        79.62000     67.72000   
8   Kep. Bangka Belitung   2022         36.38        92.47000     77.54000   
9         Kepulauan Riau   2022          5.83        88.90000     72.53000   
10           DKI Jakarta   2022          0.00        83.22000     81.12000   
11            Jawa Barat   2022         84.59        79.11000     73.71000   
12           Jawa Tengah   2022         88.88        81.47000     80.69000   
13         DI Yogyakarta   2022         84.20        79.66000     79.97000   
14            Jawa Timur   2022         90.64        79.42000     74.27000   
15                Banten   2022         82.81        83.99000     62.79000   
16                  Bali   2022         79.08        94.00000     82.08000   
17   Nusa Tenggara Barat   2022         92.85        76.24000     67.81000   
18   Nusa Tenggara Timur   2022         86.25        62.16375     60.55000   
19      Kalimantan Barat   2022         76.21        86.33000     56.29000   
20     Kalimantan Tengah   2022         49.87        91.07000     67.55000   
21    Kalimantan Selatan   2022         88.69        90.68000     69.01000   
22      Kalimantan Timur   2022         52.33        90.27000     79.43000   
23      Kalimantan Utara   2022         43.50        89.99000     72.21000   
24        Sulawesi Utara   2022         70.37        81.42000     71.42000   
25       Sulawesi Tengah   2022         77.98        80.39000     70.88000   
26      Sulawesi Selatan   2022         94.71        83.52000     71.22000   
27     Sulawesi Tenggara   2022         71.80        78.24000     74.38000   
28             Gorontalo   2022         94.25        79.76000     71.91000   
29        Sulawesi Barat   2022         76.11        83.45000     65.42000   
30                Maluku   2022         45.87        68.57000     62.27000   
31          Maluku Utara   2022         14.50        88.12000     60.61000   
32           Papua Barat   2022         11.81        62.16375     56.17000   
33                 Papua   2022         17.68        62.16375     47.88375   

    ketersediaan normalize  keterjangkauan normalize  pemanfaatan normalize  \
0                 0.844428                  0.225411               0.537084   
1                 0.844639                  0.403824               0.484154   
2                 0.921951                  0.693117               0.700552   
3                 0.408469                  0.764419               0.641189   
4                 0.615968                  0.623071               0.560186   
5                 0.859174                  0.375240               0.392039   
6                 0.736465                  0.408850               0.399057   
7                 1.000000                  0.548314               0.580071   
8                 0.383189                  0.951942               0.867237   
9                 0.061407                  0.839805               0.720730   
10                0.000000                  0.661392               0.971927   
11                0.890984                  0.532294               0.755236   
12                0.936170                  0.606423               0.959352   
13                0.886876                  0.549570               0.938297   
14                0.954708                  0.542031               0.771612   
15                0.8

Menampilkan data provinsi berdasarkan cluster pada tahun 2022.

In [ ]:
# Check the actual column names in the DataFrame
data_cluster_kmedoids.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   provinsi                  34 non-null     object 
 1   tahun                     34 non-null     int64  
 2   ketersediaan              34 non-null     float64
 3   keterjangkauan            34 non-null     float64
 4   pemanfaatan               34 non-null     float64
 5   ketersediaan normalize    34 non-null     float64
 6   keterjangkauan normalize  34 non-null     float64
 7   pemanfaatan normalize     34 non-null     float64
 8   cluster_kmeans            34 non-null     int32  
dtypes: float64(6), int32(1), int64(1), object(1)
memory usage: 2.4+ KB


In [ ]:
# Import library yang dibutuhkan
from sklearn.preprocessing import StandardScaler

# Normalisasi data menggunakan StandardScaler
scaler = StandardScaler()
features = ['ketersediaan', 'keterjangkauan', 'pemanfaatan']
normalized_data = scaler.fit_transform(data_cluster_kmedoids[features])

# Membuat DataFrame dari data yang telah dinormalisasi
normalized_df = pd.DataFrame(normalized_data, columns=[f"{col} normalize" for col in features])

# Menggabungkan kembali dengan kolom provinsi
normalized_df['provinsi'] = data_cluster_2022['provinsi']

# Menampilkan data yang telah dinormalisasi untuk verifikasi
normalized_df.head()

ketersediaan normalize  keterjangkauan normalize  pemanfaatan normalize  \
0                0.541719                 -1.353497              -0.329898   
1                0.542432                 -0.674959              -0.562896   
2                0.803919                  0.425277               0.389690   
3               -0.932797                  0.696453               0.128373   
4               -0.230986                  0.158879              -0.228203   

         provinsi  
0            Aceh  
1  Sumatera Utara  
2  Sumatera Barat  
3            Riau  
4           Jambi

Menormalisasi nilai pada fitur ketersediaan, keterjangkauan, dan pemanfaatan menggunakan *standardscaler*, kemudian menggabungkannya dengan provinsi untuk analisis lebih lanjut.

### <b>D.2. <span style='color:#3A6D8C'>Identifikasi Silhouette Score</span></b>

In [ ]:
from sklearn_extra.cluster import KMedoids
from sklearn.metrics import silhouette_score
from sklearn.metrics import davies_bouldin_score
from sklearn.preprocessing import StandardScaler

# Definisikan kolom yang akan di clustering
X = normalized_df[[
    'ketersediaan normalize',
    'keterjangkauan normalize',
    'pemanfaatan normalize'
]]


k_range = range(2, 16)
silhouette = list()

# Melakukan K-Medoids dan menghitung Silhouette Score
for k in k_range:
    kmedoids = KMedoids(n_clusters=k, random_state=0)
    cluster_labels = kmedoids.fit_predict(X)
    score = silhouette_score(X, cluster_labels)
    silhouette.append(score)

In [ ]:
import plotly.express as px

# Plot grafik
fig = px.line(
    x = k_range,
    y = silhouette,
    color_discrete_sequence = ['#FFAD60']
)

fig.update_layout(
    width = 1200,
    height = 500,
    showlegend = False,
    plot_bgcolor = 'rgba(0, 0, 0, 0)',
    title = dict(
        text = "<b>Nilai <span style='color:#FFAD60'>Silhouette Score</span>\
        </b><br><sup><sup>dengan Algoritma K-Medoids</sup></sup>",
        font = dict(
            size = 25,
            color = '#757882'
        ),
    ),
    xaxis = dict(
        showgrid = False,
        showline = True,
        linecolor = 'black',
        showticklabels = True,
        zeroline = True,
    ),
    yaxis = dict(
        showgrid = False,
        showline = True,
        linecolor='black',
        showticklabels = True,
        zeroline = True,
    ),
    margin = dict(
        l = 50
    )
)

fig.show(renderer = 'colab')

Dari hasil Silhouette Score didapatkan jumlah cluster terbaik yaitu (K) = 2 dengan nilai sebesar **0.35**

### <b>D.3. <span style='color:#3A6D8C'>Penentuan Jumlah Cluster</span></b>

In [ ]:
kmedoids = KMedoids(n_clusters = 2, random_state = 0)
cluster_labels = kmedoids.fit_predict(X)

normalized_df['cluster_kmedoids'] = cluster_labels + 1

In [ ]:
display(normalized_df)

ketersediaan normalize  keterjangkauan normalize  pemanfaatan normalize  \
0                 0.541719                 -1.353497              -0.329898   
1                 0.542432                 -0.674959              -0.562896   
2                 0.803919                  0.425277               0.389690   
3                -0.932797                  0.696453               0.128373   
4                -0.230986                  0.158879              -0.228203   
5                 0.591594                 -0.783668              -0.968388   
6                 0.176564                 -0.655845              -0.937494   
7                 1.067900                 -0.125438              -0.140668   
8                -1.018297                  1.409635               1.123439   
9                -2.106639                  0.983160               0.478513   
10               -2.314333                  0.304621               1.584285   
11                0.699181                 -0.186364               0.630412   
12                0.852012                  0.095564               1.528932   
13                0.685288                 -0.120660               1.436248   
14                0.914712                 -0.149331               0.702499   
15                0.635769                  0.396606              -0.775297   
16                0.502888                  1.592411               1.707864   
17                0.993443                 -0.529217              -0.129083   
18                0.758319                 -2.210779              -1.063647   
19                0.400644                  0.676145              -1.612028   
20               -0.537717                  1.242390              -0.162552   
21                0.845244                  1.195800               0.025391   
22               -0.450079                  1.146821               1.366735   
23               -0.764648                  1.113372               0.437320   
24                0.192595                  0.089591               0.335625   
25                0.463701                 -0.033453               0.266112   
26                1.059706                  0.340459               0.309879   
27                0.243538                 -0.290295               0.716659   
28                1.043318                 -0.108714               0.398701   
29                0.397082                  0.332097              -0.436742   
30               -0.680216                 -1.445482              -0.842235   
31               -1.797771                  0.889980              -1.055923   
32               -1.893602                 -2.210779              -1.627475   
33               -1.684484                 -2.210779              -2.694146   

                provinsi  cluster_kmedoids  
0                   Aceh                 1  
1         Sumatera Utara                 1  
2         Sumatera Barat                 1  
3                   Riau                 2  
4                  Jambi                 1  
5       Sumatera Selatan                 1  
6               Bengkulu                 1  
7                Lampung                 1  
8   Kep. Bangka Belitung                 2  
9         Kepulauan Riau                 2  
10           DKI Jakarta                 2  
11            Jawa Barat                 1  
12           Jawa Tengah                 1  
13         DI Yogyakarta                 1  
14            Jawa Timur                 1  
15                Banten                 1  
16                  Bali                 1  
17   Nusa Tenggara Barat                 1  
18   Nusa Tenggara Timur                 1  
19      Kalimantan Barat                 1  
20     Kalimantan Tengah                 2  
21    Kalimantan Selatan                 1  
22      Kalimantan Timur                 2  
23      Kalimantan Utara                 2  
24        Sulawesi Utara                 1  
25       Sulawesi Tengah                 1  
26      Sulawesi Selata

Berdasarkan output tersebut, dapat dilihat bahwa setiap provinsi berhasil dikelompokkan ke dalam 2 kluster yang berbeda, yaitu cluster 1 dan 2. Hal ini menunjukkan bahwa algoritma K-Medoids berhasil mengelompokkan provinsi-provinsi berdasarkan kesamaan karakteristiknya.

### <b>D.4. <span style='color:#3A6D8C'>Silhouette Plot</span></b>

In [ ]:
import numpy as np
import pandas as pd
import plotly.graph_objects as go
import plotly.express as px  # Import plotly.express untuk akses warna
from sklearn.metrics import silhouette_samples, silhouette_score
from sklearn_extra.cluster import KMedoids

# Menggunakan KMedoids dengan 3 cluster
kmedoids = KMedoids(n_clusters=2, random_state=0)
cluster_labels = kmedoids.fit_predict(X)

# Menambahkan kolom 'cluster_kmedoids' ke DataFrame
normalized_df['cluster_kmedoids'] = cluster_labels + 1

# Hitung silhouette score rata-rata dan nilai untuk setiap titik
silhouette_avg = silhouette_score(X, cluster_labels)
silhouette_values = silhouette_samples(X, cluster_labels)

# Siapkan data untuk visualisasi
n_clusters = len(np.unique(cluster_labels))
silhouette_data = pd.DataFrame({
    'cluster_kmedoids': cluster_labels + 1,
    'Silhouette Coefficient': silhouette_values
})

# Urutkan data berdasarkan klaster dan nilai silhouette
silhouette_data = silhouette_data.sort_values(by=['cluster_kmedoids', 'Silhouette Coefficient'], ascending=[True, True])

# Tambahkan kolom posisi untuk membantu visualisasi
silhouette_data['Posisi'] = np.arange(len(silhouette_data))

# Membuat plot menggunakan Plotly
fig = go.Figure()

# Loop untuk setiap klaster
for cluster_id in range(1, n_clusters + 1):
    cluster_data = silhouette_data[silhouette_data['cluster_kmedoids'] == cluster_id]
    fig.add_trace(go.Bar(
        x=cluster_data['Silhouette Coefficient'],
        y=cluster_data['Posisi'],
        orientation='h',
        marker=dict(color=px.colors.qualitative.Plotly[cluster_id % len(px.colors.qualitative.Plotly)]),
        name=f'Cluster {cluster_id}'
    ))

# Garis rata-rata silhouette score
fig.add_vline(
    x=silhouette_avg,
    line=dict(color='red', dash='dash'),
    annotation_text="Rata-rata Silhouette",
    annotation_position="top right"
)

# Layout
fig.update_layout(
    width=1200,
    height=600,
    title=dict(
        text="<b>Silhouette Plot</b><br><sup><sup>dengan Algoritma K-Medoids</sup></sup>",
        font=dict(size=25, color='#757882'),
    ),
    xaxis=dict(
        title="Nilai Silhouette Coefficient",
        linecolor="black",
        showgrid=False,
        zeroline=True
    ),
    yaxis=dict(
        title="Cluster",
        showticklabels=False,
        linecolor="black",
        showgrid=False
    ),
    plot_bgcolor='rgba(0, 0, 0, 0)',
    showlegend=False,
    margin=dict(l=50)
)

fig.show(renderer='colab')


Dari hasil Silhouette Plot, rata-rata Silhouette Score bernilai 0.35, hal ini menunjukan pada cluster hijau memiliki beberapa anggota dengan nilai silhouette yang mendekati 0, yang menunjukkan bahwa beberapa objek dalam cluster tersebut kurang jelas keanggotaannya atau berada dekat dengan perbatasan cluster lain. Sedangkan pada cluster merah memiliki nilai silhouette positif yang tinggi menunjukkan bahwa pengelompokan untuk cluster ini cukup baik.



### <b>D.5. <span style='color:#3A6D8C'>Plot Hasil Cluster</span></b>

In [ ]:
import plotly.graph_objects as go

# Membuat plot 3D
fig = go.Figure(
    data = [
        go.Scatter3d(
            x = data_cluster_2022['ketersediaan'],
            y = data_cluster_2022['keterjangkauan'],
            z = data_cluster_2022['pemanfaatan'],
            mode = 'markers',
            marker = dict(
                size = 5,
                color = normalized_df['cluster_kmedoids'],
                opacity = 0.5,
                colorscale='Viridis',
                colorbar = dict(title='Color Legend')
            ),
            hovertext = normalized_df['provinsi'],
            hovertemplate='<b>%{hovertext}</b><br>ketersediaan: %{x}<br>Keterjangkauan: %{y}<br>Pemanfaatan: %{z}<br>Cluster: %{marker.color}'
        )
    ]
)


fig.update_layout(
    width = 800,
    height = 800,
    plot_bgcolor = 'rgba(0, 0, 0, 0)',
    title = dict(
        text = '',
        font = dict(
            size = 20
        )
    ),
    scene = dict(
        xaxis_title = 'Ketersediaan',
        yaxis_title = 'Keterjangkauaan',
        zaxis_title = 'Pemanfaatan',
        xaxis=dict(showgrid=False),
        yaxis=dict(showgrid=False),
        zaxis=dict(showgrid=False)
    )
)

# Tampilkan plot
fig.show()

Proses clustering ini menghasilkan 2 kelompok (cluster) berbeda berdasarkan karakteristik kesamaan dalam kedua aspek tersebut. Berdasarkan hasil clustering, berikut adalah pembagian provinsi dalam masing-masing cluster :

Cluster 1 (Ungu): keterjangkauan tinggi dan pemanfaatan optimal, tetapi memiliki ketersediaan yang tidak selalu konsisten.

Cluster 2 (Kuning): ketersediaan tinggi, tetapi keterjangkauan lebih rendah, meskipun pemanfaatannya tetap tinggi.

### <b>D.6. <span style='color:#3A6D8C'>Summary</span></b>

In [ ]:
# Add the 'cluster_kmedoids' column to 'data_cluster_kmedoids'
data_cluster_kmedoids['cluster_kmedoids'] = normalized_df['cluster_kmedoids']

# Menghitung mean dan median untuk setiap cluster
cluster_statistics_kmedoids = data_cluster_kmedoids.groupby('cluster_kmedoids').agg({
    'ketersediaan': ['mean', 'median', 'std'],
    'keterjangkauan': ['mean', 'median', 'std'],
    'pemanfaatan': ['mean', 'median', 'std']
}).reset_index()

# Menampilkan hasil
display(cluster_statistics_kmedoids)

cluster_kmedoids ketersediaan                   keterjangkauan         \
                           mean median        std           mean median   
0                1     80.75375  82.19  11.666728      79.496406  79.71   
1                2     27.06800  27.03  19.190912      83.486750  88.51   

             pemanfaatan                    
         std        mean median        std  
0   6.834371   68.949167  68.41   6.652535  
1  11.526641   68.485375  71.01  10.710240

**Cluster 1 (Ungu)**

Ketersediaan

Cluster ini memiliki ketersediaan yang cenderung tinggi, dengan rata-rata mencapai 80.75 dan median sedikit lebih tinggi di 82.19. Ini menunjukkan bahwa mayoritas entitas atau objek dalam cluster ini tersedia secara relatif baik.

Keterjangkauan

Cluster ini juga memiliki keterjangkauan yang tinggi, dengan rata-rata 79.50 dan median yang sangat dekat pada 79.71. Ini menunjukkan bahwa mayoritas objek atau entitas dalam cluster ini mudah dijangkau atau diakses.

Pemanfaatan

Cluster ini memiliki pemanfaatan pada tingkat moderat dengan nilai rata-rata 68.95 dan median 68.41, yang menunjukkan bahwa meskipun objek atau entitas diakses dengan baik, tingkat penggunaannya tidak maksimal.

**Cluster 2 (Kuning)**

Ketersediaan

Cluster ini memiliki ketersediaan yang sangat rendah, dengan rata-rata hanya 27.07 dan median yang hampir sama di 27.03. Ini menunjukkan bahwa objek atau entitas dalam cluster ini terbatas dalam ketersediaannya, mungkin sulit ditemukan atau hanya tersedia dalam jumlah sedikit.

Keterjangkauan

Cluster ini memiliki keterjangkauan yang sangat tinggi, dengan rata-rata 83.49 dan median 88.51, menunjukkan bahwa meskipun ketersediaannya rendah, aksesibilitas terhadap objek atau entitas ini sangat baik.

Pemanfaatan

Cluster ini memiliki pemanfaatan pada tingkat moderat, sedikit lebih tinggi dibandingkan dengan cluster 1, dengan rata-rata 68.49 dan median 71.01. Ini menunjukkan bahwa meskipun objek atau entitas dalam cluster ini tidak terlalu sering tersedia, mereka cukup sering digunakan atau dimanfaatkan.

#<b>E. <span style='color:#3A6D8C'>Modelling dengan Agglomerative Clustering</span></b>

Proses modelling dilakukan menggunakan K-Medoids. K-Medoids ialah algoritma clustering yang hampir sama dengan algoritma K-Means. Perbedaannya adalah K-Medoids mengunakan objek sebagai pusat cluster untuk setiap cluster, sedangkan K-Means menggunakan nilai rata-rata sebagai pusat cluster untuk setiap cluster.

In [ ]:
data_cluster_agglomerative = data_cluster_kmedoids.copy()
display(data_cluster_agglomerative)

provinsi  tahun  ketersediaan  keterjangkauan  pemanfaatan  \
0                   Aceh   2022         80.17        69.34000     66.25000   
1         Sumatera Utara   2022         80.19        75.02000     64.44000   
2         Sumatera Barat   2022         87.53        84.23000     71.84000   
3                   Riau   2022         38.78        86.50000     69.81000   
4                  Jambi   2022         58.48        82.00000     67.04000   
5       Sumatera Selatan   2022         81.57        74.11000     61.29000   
6               Bengkulu   2022         69.92        75.18000     61.53000   
7                Lampung   2022         94.94        79.62000     67.72000   
8   Kep. Bangka Belitung   2022         36.38        92.47000     77.54000   
9         Kepulauan Riau   2022          5.83        88.90000     72.53000   
10           DKI Jakarta   2022          0.00        83.22000     81.12000   
11            Jawa Barat   2022         84.59        79.11000     73.71000   
12           Jawa Tengah   2022         88.88        81.47000     80.69000   
13         DI Yogyakarta   2022         84.20        79.66000     79.97000   
14            Jawa Timur   2022         90.64        79.42000     74.27000   
15                Banten   2022         82.81        83.99000     62.79000   
16                  Bali   2022         79.08        94.00000     82.08000   
17   Nusa Tenggara Barat   2022         92.85        76.24000     67.81000   
18   Nusa Tenggara Timur   2022         86.25        62.16375     60.55000   
19      Kalimantan Barat   2022         76.21        86.33000     56.29000   
20     Kalimantan Tengah   2022         49.87        91.07000     67.55000   
21    Kalimantan Selatan   2022         88.69        90.68000     69.01000   
22      Kalimantan Timur   2022         52.33        90.27000     79.43000   
23      Kalimantan Utara   2022         43.50        89.99000     72.21000   
24        Sulawesi Utara   2022         70.37        81.42000     71.42000   
25       Sulawesi Tengah   2022         77.98        80.39000     70.88000   
26      Sulawesi Selatan   2022         94.71        83.52000     71.22000   
27     Sulawesi Tenggara   2022         71.80        78.24000     74.38000   
28             Gorontalo   2022         94.25        79.76000     71.91000   
29        Sulawesi Barat   2022         76.11        83.45000     65.42000   
30                Maluku   2022         45.87        68.57000     62.27000   
31          Maluku Utara   2022         14.50        88.12000     60.61000   
32           Papua Barat   2022         11.81        62.16375     56.17000   
33                 Papua   2022         17.68        62.16375     47.88375   

    ketersediaan normalize  keterjangkauan normalize  pemanfaatan normalize  \
0                 0.844428                  0.225411               0.537084   
1                 0.844639                  0.403824               0.484154   
2                 0.921951                  0.693117               0.700552   
3                 0.408469                  0.764419               0.641189   
4                 0.615968                  0.623071               0.560186   
5                 0.859174                  0.375240               0.392039   
6                 0.736465                  0.408850               0.399057   
7                 1.000000                  0.548314               0.580071   
8                 0.383189                  0.951942               0.867237   
9                 0.061407                  0.839805               0.720730   
10                0.000000                  0.661392               0.971927   
11                0.890984                  0.532294               0.755236   
12                0.936170                  0.606423               0.959352   
13                0.886876                  0.549570               0.938297   
14                0.954708                  0.542031               0.771612   
15                0.8

Menampilkan data provinsi berdasarkan cluster pada tahun 2022 dengan algoritma K-Means dan K-Medoids.

In [ ]:
# Menghapus kolom normalize dan cluster yang tidak diperlukan
data_cluster_agglomerative = data_cluster_agglomerative.drop(columns=['ketersediaan normalize', 'keterjangkauan normalize', 'pemanfaatan normalize', 'cluster_kmeans', 'cluster_kmedoids'])

# Menampilkan hasil setelah penghapusan kolom
display(data_cluster_agglomerative.head())


provinsi  tahun  ketersediaan  keterjangkauan  pemanfaatan
0            Aceh   2022         80.17           69.34        66.25
1  Sumatera Utara   2022         80.19           75.02        64.44
2  Sumatera Barat   2022         87.53           84.23        71.84
3            Riau   2022         38.78           86.50        69.81
4           Jambi   2022         58.48           82.00        67.04

Menghapus kolom yang tidak dibutuhkan

### <b>E.1. <span style='color:#3A6D8C'>Proses Scalling</span></b>

In [ ]:
# Import library yang dibutuhkan
from sklearn.preprocessing import StandardScaler

# Normalisasi data menggunakan StandardScaler
scaler = StandardScaler()
features = ['ketersediaan', 'keterjangkauan', 'pemanfaatan']
normalized_data = scaler.fit_transform(data_cluster_kmedoids[features])

# Membuat DataFrame dari data yang telah dinormalisasi
normalized_df = pd.DataFrame(normalized_data, columns=[f"{col} normalize" for col in features])

# Menggabungkan kembali dengan kolom provinsi
normalized_df['provinsi'] = data_cluster_2022['provinsi']

# Menampilkan data yang telah dinormalisasi untuk verifikasi
normalized_df.head()

ketersediaan normalize  keterjangkauan normalize  pemanfaatan normalize  \
0                0.541719                 -1.353497              -0.329898   
1                0.542432                 -0.674959              -0.562896   
2                0.803919                  0.425277               0.389690   
3               -0.932797                  0.696453               0.128373   
4               -0.230986                  0.158879              -0.228203   

         provinsi  
0            Aceh  
1  Sumatera Utara  
2  Sumatera Barat  
3            Riau  
4           Jambi

Menormalisasi nilai pada fitur ketersediaan, keterjangkauan, dan pemanfaatan menggunakan *standardscaler*, kemudian menggabungkannya dengan provinsi untuk analisis lebih lanjut.

In [ ]:
# Proses masukan data
for col in normalized_df.columns:
    data_cluster_agglomerative[col] = normalized_df[col]

# Tampilkan hasilnya
display(data_cluster_agglomerative)

provinsi  tahun  ketersediaan  keterjangkauan  pemanfaatan  \
0                   Aceh   2022         80.17        69.34000     66.25000   
1         Sumatera Utara   2022         80.19        75.02000     64.44000   
2         Sumatera Barat   2022         87.53        84.23000     71.84000   
3                   Riau   2022         38.78        86.50000     69.81000   
4                  Jambi   2022         58.48        82.00000     67.04000   
5       Sumatera Selatan   2022         81.57        74.11000     61.29000   
6               Bengkulu   2022         69.92        75.18000     61.53000   
7                Lampung   2022         94.94        79.62000     67.72000   
8   Kep. Bangka Belitung   2022         36.38        92.47000     77.54000   
9         Kepulauan Riau   2022          5.83        88.90000     72.53000   
10           DKI Jakarta   2022          0.00        83.22000     81.12000   
11            Jawa Barat   2022         84.59        79.11000     73.71000   
12           Jawa Tengah   2022         88.88        81.47000     80.69000   
13         DI Yogyakarta   2022         84.20        79.66000     79.97000   
14            Jawa Timur   2022         90.64        79.42000     74.27000   
15                Banten   2022         82.81        83.99000     62.79000   
16                  Bali   2022         79.08        94.00000     82.08000   
17   Nusa Tenggara Barat   2022         92.85        76.24000     67.81000   
18   Nusa Tenggara Timur   2022         86.25        62.16375     60.55000   
19      Kalimantan Barat   2022         76.21        86.33000     56.29000   
20     Kalimantan Tengah   2022         49.87        91.07000     67.55000   
21    Kalimantan Selatan   2022         88.69        90.68000     69.01000   
22      Kalimantan Timur   2022         52.33        90.27000     79.43000   
23      Kalimantan Utara   2022         43.50        89.99000     72.21000   
24        Sulawesi Utara   2022         70.37        81.42000     71.42000   
25       Sulawesi Tengah   2022         77.98        80.39000     70.88000   
26      Sulawesi Selatan   2022         94.71        83.52000     71.22000   
27     Sulawesi Tenggara   2022         71.80        78.24000     74.38000   
28             Gorontalo   2022         94.25        79.76000     71.91000   
29        Sulawesi Barat   2022         76.11        83.45000     65.42000   
30                Maluku   2022         45.87        68.57000     62.27000   
31          Maluku Utara   2022         14.50        88.12000     60.61000   
32           Papua Barat   2022         11.81        62.16375     56.17000   
33                 Papua   2022         17.68        62.16375     47.88375   

    ketersediaan normalize  keterjangkauan normalize  pemanfaatan normalize  
0                 0.541719                 -1.353497              -0.329898  
1                 0.542432                 -0.674959              -0.562896  
2                 0.803919                  0.425277               0.389690  
3                -0.932797                  0.696453               0.128373  
4                -0.230986                  0.158879              -0.228203  
5                 0.591594                 -0.783668              -0.968388  
6                 0.176564                 -0.655845              -0.937494  
7                 1.067900                 -0.125438              -0.140668  
8                -1.018297                  1.409635               1.123439  
9                -2.106639                  0.983160               0.478513  
10               -2.314333                  0.304621               1.584285  
11                0.699181                 -0.186364               0.630412  
12                0.852012                  0.095564               1.528932  
13                0.685288                 -0.120660               1.436248  
14                0.914712                 -0.149331               0.702499  
15                0.635769           

Memasukkan data dari DataFrame normalized_df ke DataFrame data_cluster_agglomerative

### <b>E.2. <span style='color:#3A6D8C'>Identifikasi Silhouette Score</span></b>

In [ ]:
from sklearn.cluster import AgglomerativeClustering
import numpy as np
import matplotlib.pyplot as plt
import scipy.cluster.hierarchy as sch
from sklearn.metrics import silhouette_score

# Definisikan kolom yang akan di clustering
X = data_cluster_agglomerative[[
    'ketersediaan normalize',
    'keterjangkauan normalize',
    'pemanfaatan normalize'
]]

k_range = range(2, 16)
silhouette = list()

# Melakukan AgglomerativeClustering dan menghitung Davies-Bouldin Index & Calinski-Harabasz Index
for k in k_range:
    agg_clust = AgglomerativeClustering(n_clusters=k, linkage='ward')
    cluster_labels = agg_clust.fit_predict(X)  # Gunakan X_scaled sesuai data yang sudah dinormalisasi

    #Silhoutte score
    score = silhouette_score(X, cluster_labels)
    silhouette.append(score)

In [ ]:
import plotly.express as px

# Plot grafik
fig = px.line(
    x = k_range,
    y = silhouette,
    color_discrete_sequence = ['#FFAD60']
)

fig.update_layout(
    width = 1200,
    height = 500,
    showlegend = False,
    plot_bgcolor = 'rgba(0, 0, 0, 0)',
    title = dict(
        text = "<b>Nilai <span style='color:#FFAD60'>Silhouette Score</span>\
        </b><br><sup><sup>dengan Algoritma Agglomerative</sup></sup>",
        font = dict(
            size = 25,
            color = '#757882'
        ),
    ),
    xaxis = dict(
        showgrid = False,
        showline = True,
        linecolor = 'black',
        showticklabels = True,
        zeroline = True,
    ),
    yaxis = dict(
        showgrid = False,
        showline = True,
        linecolor='black',
        showticklabels = True,
        zeroline = True,
    ),
    margin = dict(
        l = 50
    )
)

fig.show(renderer = 'colab')

Dari hasil Silhouette Score didapatkan jumlah cluster terbaik yaitu (K) = 4 dengan nilai sebesar **0.36**

### <b>E.3. <span style='color:#3A6D8C'>Penentuan Jumlah Cluster</span></b>

In [ ]:
from sklearn.cluster import AgglomerativeClustering

# Menggunakan AgglomerativeClustering
agg_clust = AgglomerativeClustering(n_clusters=4, linkage='ward')
cluster_labels = agg_clust.fit_predict(X)

# Menambahkan label cluster ke dalam DataFrame
data_cluster_agglomerative['cluster_agglomerative'] = cluster_labels + 1  # Menambahkan 1 agar label mulai dari 1

In [ ]:
display(data_cluster_agglomerative)

provinsi  tahun  ketersediaan  keterjangkauan  pemanfaatan  \
0                   Aceh   2022         80.17        69.34000     66.25000   
1         Sumatera Utara   2022         80.19        75.02000     64.44000   
2         Sumatera Barat   2022         87.53        84.23000     71.84000   
3                   Riau   2022         38.78        86.50000     69.81000   
4                  Jambi   2022         58.48        82.00000     67.04000   
5       Sumatera Selatan   2022         81.57        74.11000     61.29000   
6               Bengkulu   2022         69.92        75.18000     61.53000   
7                Lampung   2022         94.94        79.62000     67.72000   
8   Kep. Bangka Belitung   2022         36.38        92.47000     77.54000   
9         Kepulauan Riau   2022          5.83        88.90000     72.53000   
10           DKI Jakarta   2022          0.00        83.22000     81.12000   
11            Jawa Barat   2022         84.59        79.11000     73.71000   
12           Jawa Tengah   2022         88.88        81.47000     80.69000   
13         DI Yogyakarta   2022         84.20        79.66000     79.97000   
14            Jawa Timur   2022         90.64        79.42000     74.27000   
15                Banten   2022         82.81        83.99000     62.79000   
16                  Bali   2022         79.08        94.00000     82.08000   
17   Nusa Tenggara Barat   2022         92.85        76.24000     67.81000   
18   Nusa Tenggara Timur   2022         86.25        62.16375     60.55000   
19      Kalimantan Barat   2022         76.21        86.33000     56.29000   
20     Kalimantan Tengah   2022         49.87        91.07000     67.55000   
21    Kalimantan Selatan   2022         88.69        90.68000     69.01000   
22      Kalimantan Timur   2022         52.33        90.27000     79.43000   
23      Kalimantan Utara   2022         43.50        89.99000     72.21000   
24        Sulawesi Utara   2022         70.37        81.42000     71.42000   
25       Sulawesi Tengah   2022         77.98        80.39000     70.88000   
26      Sulawesi Selatan   2022         94.71        83.52000     71.22000   
27     Sulawesi Tenggara   2022         71.80        78.24000     74.38000   
28             Gorontalo   2022         94.25        79.76000     71.91000   
29        Sulawesi Barat   2022         76.11        83.45000     65.42000   
30                Maluku   2022         45.87        68.57000     62.27000   
31          Maluku Utara   2022         14.50        88.12000     60.61000   
32           Papua Barat   2022         11.81        62.16375     56.17000   
33                 Papua   2022         17.68        62.16375     47.88375   

    ketersediaan normalize  keterjangkauan normalize  pemanfaatan normalize  \
0                 0.541719                 -1.353497              -0.329898   
1                 0.542432                 -0.674959              -0.562896   
2                 0.803919                  0.425277               0.389690   
3                -0.932797                  0.696453               0.128373   
4                -0.230986                  0.158879              -0.228203   
5                 0.591594                 -0.783668              -0.968388   
6                 0.176564                 -0.655845              -0.937494   
7                 1.067900                 -0.125438              -0.140668   
8                -1.018297                  1.409635               1.123439   
9                -2.106639                  0.983160               0.478513   
10               -2.314333                  0.304621               1.584285   
11                0.699181                 -0.186364               0.630412   
12                0.852012                  0.095564               1.528932   
13                0.685288                 -0.120660               1.436248   
14                0.914712                 -0.149331               0.702499   
15                0.6

Berdasarkan output tersebut, dapat dilihat bahwa setiap provinsi berhasil dikelompokkan ke dalam 4 kluster yang berbeda, yaitu cluster 1,2,3, dan 4. Hal ini menunjukkan bahwa algoritma Agglomerative Clustering berhasil mengelompokkan provinsi-provinsi berdasarkan kesamaan karakteristiknya.

### <b>E.4. <span style='color:#3A6D8C'>Silhouette Plot</span></b>

In [ ]:
import numpy as np
import pandas as pd
import plotly.graph_objects as go
import plotly.express as px  # Import plotly.express untuk akses warna
from sklearn.metrics import silhouette_samples, silhouette_score
from sklearn.cluster import AgglomerativeClustering

# Menggunakan AgglomerativeClustering
agg_clust = AgglomerativeClustering(n_clusters=4, linkage='ward')
data_cluster_agglomerative['cluster_agglomerative'] = agg_clust.fit_predict(X)

# Hitung silhouette score rata-rata dan nilai untuk setiap titik
silhouette_avg = silhouette_score(X, data_cluster_agglomerative['cluster_agglomerative'])
silhouette_values = silhouette_samples(X, data_cluster_agglomerative['cluster_agglomerative'])

# Siapkan data untuk visualisasi
n_clusters = len(np.unique(data_cluster_agglomerative['cluster_agglomerative']))
silhouette_data = pd.DataFrame({
    'cluster_agglomerative': data_cluster_agglomerative['cluster_agglomerative'] + 1,
    'Silhouette Coefficient': silhouette_values
})

# Urutkan data berdasarkan klaster dan nilai silhouette
silhouette_data = silhouette_data.sort_values(by=['cluster_agglomerative', 'Silhouette Coefficient'], ascending=[True, True])

# Tambahkan kolom posisi untuk membantu visualisasi
silhouette_data['Posisi'] = np.arange(len(silhouette_data))

# Membuat plot menggunakan Plotly
fig = go.Figure()

# Loop untuk setiap klaster
for cluster_id in range(1, n_clusters + 1):
    cluster_data = silhouette_data[silhouette_data['cluster_agglomerative'] == cluster_id]
    fig.add_trace(go.Bar(
        x=cluster_data['Silhouette Coefficient'],
        y=cluster_data['Posisi'],
        orientation='h',
        marker=dict(color=px.colors.qualitative.Plotly[cluster_id % len(px.colors.qualitative.Plotly)]),
        name=f'Cluster {cluster_id}'
    ))

# Garis rata-rata silhouette score
fig.add_vline(
    x=silhouette_avg,
    line=dict(color='red', dash='dash'),
    annotation_text="Rata-rata Silhouette",
    annotation_position="top right"
)

# Layout
fig.update_layout(
    width=1200,
    height=600,
    title=dict(
        text="<b>Silhouette Plot</b><br><sup><sup>dengan Algoritma Agglomerative</sup></sup>",
        font=dict(size=25, color='#757882'),
    ),
    xaxis=dict(
        title="Nilai Silhouette Coefficient",
        linecolor="black",
        showgrid=False,
        zeroline=True
    ),
    yaxis=dict(
        title="Cluster",
        showticklabels=False,
        linecolor="black",
        showgrid=False
    ),
    plot_bgcolor='rgba(0, 0, 0, 0)',
    showlegend=False,
    margin=dict(l=50)
)

fig.show(renderer='colab')


Dari hasil Silhouette Plot, rata-rata Silhouette Score bernilai 0.36, hal ini menunjukan :

cluster **orange** memiliki nilai Silhouette Coefficient yang sangat tinggi, dengan banyak batang mendekati atau lebih dari 0.6.
Ini menunjukkan bahwa kluster oranye memiliki pengelompokan yang sangat baik, di mana objek-objek dalam kluster ini secara jelas berada di kluster yang tepat.


Pada cluster **Ungu** menunjukkan kluster dengan nilai Silhouette Coefficient yang sebagian besar positif tetapi bervariasi, ada yang lebih rendah mendekati 0.Hal ini menunjukkan bahwa beberapa objek dalam kluster ini terklasifikasi dengan baik, namun ada juga objek yang mungkin berada di tepi antar kluster.

Pada cluster **Hijau** memiliki nilai Silhouette yang cukup tinggi dan konsisten, dengan sebagian besar batang berada di sekitar nilai rata-rata atau lebih tinggi.Ini menandakan bahwa kluster hijau memiliki pengelompokan yang cukup baik dan sedikit objek yang berada di posisi ambigu.

Pada cluster **Merah** menunjukkan kluster dengan distribusi nilai yang sangat bervariasi, dari mendekati 0 hingga positif yang lebih tinggi. Ini menandakan bahwa kluster merah mungkin memiliki beberapa objek yang terklasifikasi dengan baik, namun juga ada beberapa objek yang mungkin kurang cocok dengan kluster ini atau berada di perbatasan kluster.




### <b>E.5. <span style='color:#3A6D8C'>Plot Hasil Cluster</span></b>

In [ ]:
# Membuat urutan label cluster dimulai dari 1
cluster_labels = data_cluster_agglomerative['cluster_agglomerative']
cluster_labels = cluster_labels - cluster_labels.min() + 1  # Pastikan mulai dari 1

# Membuat plot 3D
fig = go.Figure(
    data = [
        go.Scatter3d(
            x = data_cluster_agglomerative['ketersediaan'],
            y = data_cluster_agglomerative['keterjangkauan'],
            z = data_cluster_agglomerative['pemanfaatan'],
            mode = 'markers',
            marker = dict(
                size = 5,
                color = cluster_labels,  # Gunakan cluster_labels yang sudah dimodifikasi
                opacity = 0.5,
                colorscale='Jet',
                colorbar = dict(title='Color Legend')
            ),
            hovertext = data_cluster_agglomerative['provinsi'],
            hovertemplate='<b>%{hovertext}</b><br>ketersediaan: %{x}<br>Keterjangkauan: %{y}<br>Pemanfaatan: %{z}<br>Cluster: %{marker.color}'
        )
    ]
)

fig.update_layout(
    width = 800,
    height = 800,
    plot_bgcolor = 'rgba(0, 0, 0, 0)',
    title = dict(
        text = '',
        font = dict(
            size = 20
        )
    ),
    scene = dict(
        xaxis_title = 'Ketersediaan',
        yaxis_title = 'Keterjangkauaan',
        zaxis_title = 'Pemanfaatan',
        xaxis=dict(showgrid=False),
        yaxis=dict(showgrid=False),
        zaxis=dict(showgrid=False)
    )
)

# Tampilkan plot
fig.show()


Proses clustering ini menghasilkan 4 kelompok (cluster) berbeda berdasarkan karakteristik kesamaan dalam kedua aspek tersebut. Berdasarkan hasil clustering, berikut adalah pembagian provinsi dalam masing-masing cluster :

Cluster 1 (Ungu): Pemanfaatan perlu ditingkatkan meskipun ketersediaan dan keterjangkauan baik.

Cluster 2 (Biru Muda): Ketersediaan tinggi, pemanfaatan menengah. Perlu strategi promosi untuk meningkatkan pemanfaatan.

Cluster 3 (Kuning): Ketersediaan rendah, meskipun terjangkau dan dimanfaatkan cukup baik.

Cluster 4 (Merah): Perlu perbaikan besar di semua aspek.

### <b>E.6. <span style='color:#3A6D8C'>Summary</span></b>

In [ ]:
# Mengubah nilai cluster untuk mulai dari 1
data_cluster_agglomerative['cluster_agglomerative'] = data_cluster_agglomerative['cluster_agglomerative'] + 1

# Menghitung mean, median, dan standar deviasi untuk setiap cluster
cluster_statistics_agglomerative = data_cluster_agglomerative.groupby('cluster_agglomerative').agg({
    'ketersediaan': ['mean', 'median', 'std'],
    'keterjangkauan': ['mean', 'median', 'std'],
    'pemanfaatan': ['mean', 'median', 'std']
}).reset_index()

# Menampilkan hasil
display(cluster_statistics_agglomerative)


cluster_agglomerative ketersediaan                    keterjangkauan  \
                                mean  median        std           mean   
0                     1    83.237647  84.590  10.176940      81.737059   
1                     2    73.995000  80.180  14.776302      70.730625   
2                     3    35.585556  38.780  25.114707      89.393333   
3                     4    14.745000  14.745   4.150717      62.163750   

                      pemanfaatan                       
     median       std        mean     median       std  
0  81.42000  3.427395   70.374706  71.220000  5.886665  
1  71.72500  5.091352   62.721667  61.900000  2.181361  
2  89.99000  3.220427   73.653333  72.530000  7.077530  
3  62.16375  0.000000   52.026875  52.026875  5.859264

**Cluster 1 (Ungu)**

Ketersediaan

Cluster ini memiliki ketersediaan yang cukup tinggi, dengan nilai mean 83.24 dan median 84.59 yang menunjukkan sumber daya atau fasilitas yang tersedia cukup baik dan konsisten. Variasi yang ada tidak terlalu besar, dengan deviasi standar 10.18.

Keterjangkauan

Cluster ini juga memiliki keterjangkauan yang cukup baik dan konsisten, dengan mean 81.74 dan median 81.42 yang menunjukkan bahwa sumber daya atau fasilitas cukup terjangkau secara merata di seluruh unit dalam cluster ini. Deviasi standar 3.43 mengindikasikan konsistensi yang baik.

Pemanfaatan

Cluster ini memiliki pemanfaatan yang cukup baik, dengan mean 70.37 dan median 71.22 yang menunjukkan bahwa sumber daya atau fasilitas dimanfaatkan dengan cukup baik. Namun, deviasi standar 5.89 menunjukkan adanya variasi dalam pemanfaatan di antara unit-unit dalam cluster ini.

**Cluster 2 (Biru Muda)**

Ketersediaan

Cluster ini memiliki ketersediaan lebih rendah dibandingkan dengan Cluster 1, dengan nilai mean 73.99 dan median 80.18. Meskipun median lebih tinggi, deviasi standar 14.78 menunjukkan adanya variasi yang lebih besar dalam ketersediaan di antara unit-unit dalam cluster ini.

Keterjangkauan

Cluster ini memiliki keterjangkauan sedikit lebih rendah dibandingkan Cluster 1, dengan mean 70.73 dan median 71.73. Deviasi standar 5.09 menunjukkan adanya variasi dalam keterjangkauan di antara unit-unit dalam cluster ini.

Pemanfaatan

Cluster ini memiliki pemanfaatan yang lebih rendah dibandingkan Cluster 0, dengan mean 62.72 dan median 61.90. Deviasi standar 2.18 menunjukkan bahwa meskipun pemanfaatan rendah, konsistensinya cukup baik di seluruh unit dalam cluster ini.

**Cluster 3 (Kuning)**

Ketersediaan

Cluster ini menunjukkan ketersediaan yang sangat rendah, dengan mean hanya 35.59 dan median 38.78. Variasi sangat besar, seperti yang terlihat dari deviasi standar yang mencapai 25.11, menunjukkan ketidakseimbangan besar dalam ketersediaan sumber daya.

Keterjangkauan

Cluster ini menunjukkan keterjangkauan yang sangat baik, dengan mean 89.39 dan median 89.99, serta deviasi standar 3.22 menunjukkan bahwa fasilitas atau sumber daya sangat terjangkau dan konsisten di seluruh unit dalam cluster ini.

Pemanfaatan

Cluster ini memiliki pemanfatan yang cukup baik, dengan mean 73.65 dan median 72.53 yang menunjukkan tingkat pemanfaatan yang relatif tinggi. Namun, deviasi standar 7.08 menunjukkan adanya variasi yang signifikan dalam pemanfaatan antara unit-unit dalam cluster ini.

**Cluster 4 (Merah)**

Ketersediaan

Cluster ini memiliki ketersediaan yang sangat terbatas, dengan nilai mean dan median yang sangat rendah 14.75, serta deviasi standar 4.15, menunjukkan bahwa ketersediaan sangat terbatas dan hampir seragam di seluruh unit.

Keterjangkauan

Cluster ini memiliki keterjangkauan sangat rendah, dengan nilai mean dan median yang seragam di 62.16, serta deviasi standar 0.00 menunjukkan keterjangkauan yang terbatas dan konsisten di seluruh unit, namun cukup rendah.

Pemanfaatan

Cluster ini memiliki pemanfaatan yang sangat rendah, dengan mean dan median yang rendah 52.03 serta deviasi standar 5.86 menunjukkan bahwa pemanfaatan di cluster ini sangat terbatas dan bervariasi.

#<b>F. <span style='color:#3A6D8C'>Final Model</span></b>

Setelah melakukan berbagai percobaan algoritma, diperoleh model terbaik menggunakan K-Means dengan nilai Silhoette Score **0.3791**

### **F.1 K-MEANS**

### **F.1.1 Indentifikasi Silhouette Score**

In [ ]:
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score

# Definisikan kolom yang akan di clustering
X = data_cluster_2022[[
    'ketersediaan normalize',
    'keterjangkauan normalize',
    'pemanfaatan normalize'
]]

k_range = range(2, 16)
silhouette = list()

# Melakukan K-Means dan menghitung Silhouette Score
for k in k_range:
    kmeans = KMeans(n_clusters=k, random_state=0)
    cluster_labels = kmeans.fit_predict(X)
    score = silhouette_score(X, cluster_labels)
    silhouette.append(score)

In [ ]:
import plotly.express as px

# Plot grafik
fig = px.line(
    x = k_range,
    y = silhouette,
    color_discrete_sequence = ['#FFAD60']
)

fig.update_layout(
    width = 1200,
    height = 500,
    showlegend = False,
    plot_bgcolor = 'rgba(0, 0, 0, 0)',
    title = dict(
        text = "<b>Nilai <span style='color:#FFAD60'>Silhouette Score</span>\
        </b><br><sup><sup>dengan Algoritma K-Means</sup></sup>",
        font = dict(
            size = 25,
            color = '#757882'
        ),
    ),
    xaxis = dict(
        showgrid = False,
        showline = True,
        linecolor = 'black',
        showticklabels = True,
        zeroline = True,
    ),
    yaxis = dict(
        showgrid = False,
        showline = True,
        linecolor='black',
        showticklabels = True,
        zeroline = True,
    ),
    margin = dict(
        l = 50
    )
)

fig.show(renderer = 'colab')

Dari hasil Silhouette Score didapatkan jumlah cluster terbaik yaitu (K) = 4 dengan nilai sebesar **0.3791**

### **F.1.2 Penentuan Jumlah Cluster**

In [ ]:
kmeans = KMeans(n_clusters = 4, random_state = 0)
cluster_labels = kmeans.fit_predict(X)

data_cluster_2022['cluster_kmeans'] = cluster_labels + 1
display(data_cluster_2022)

provinsi  tahun  ketersediaan  keterjangkauan  pemanfaatan  \
0                   Aceh   2022         80.17        69.34000     66.25000   
1         Sumatera Utara   2022         80.19        75.02000     64.44000   
2         Sumatera Barat   2022         87.53        84.23000     71.84000   
3                   Riau   2022         38.78        86.50000     69.81000   
4                  Jambi   2022         58.48        82.00000     67.04000   
5       Sumatera Selatan   2022         81.57        74.11000     61.29000   
6               Bengkulu   2022         69.92        75.18000     61.53000   
7                Lampung   2022         94.94        79.62000     67.72000   
8   Kep. Bangka Belitung   2022         36.38        92.47000     77.54000   
9         Kepulauan Riau   2022          5.83        88.90000     72.53000   
10           DKI Jakarta   2022          0.00        83.22000     81.12000   
11            Jawa Barat   2022         84.59        79.11000     73.71000   
12           Jawa Tengah   2022         88.88        81.47000     80.69000   
13         DI Yogyakarta   2022         84.20        79.66000     79.97000   
14            Jawa Timur   2022         90.64        79.42000     74.27000   
15                Banten   2022         82.81        83.99000     62.79000   
16                  Bali   2022         79.08        94.00000     82.08000   
17   Nusa Tenggara Barat   2022         92.85        76.24000     67.81000   
18   Nusa Tenggara Timur   2022         86.25        62.16375     60.55000   
19      Kalimantan Barat   2022         76.21        86.33000     56.29000   
20     Kalimantan Tengah   2022         49.87        91.07000     67.55000   
21    Kalimantan Selatan   2022         88.69        90.68000     69.01000   
22      Kalimantan Timur   2022         52.33        90.27000     79.43000   
23      Kalimantan Utara   2022         43.50        89.99000     72.21000   
24        Sulawesi Utara   2022         70.37        81.42000     71.42000   
25       Sulawesi Tengah   2022         77.98        80.39000     70.88000   
26      Sulawesi Selatan   2022         94.71        83.52000     71.22000   
27     Sulawesi Tenggara   2022         71.80        78.24000     74.38000   
28             Gorontalo   2022         94.25        79.76000     71.91000   
29        Sulawesi Barat   2022         76.11        83.45000     65.42000   
30                Maluku   2022         45.87        68.57000     62.27000   
31          Maluku Utara   2022         14.50        88.12000     60.61000   
32           Papua Barat   2022         11.81        62.16375     56.17000   
33                 Papua   2022         17.68        62.16375     47.88375   

    ketersediaan normalize  keterjangkauan normalize  pemanfaatan normalize  \
0                 0.844428                  0.225411               0.537084   
1                 0.844639                  0.403824               0.484154   
2                 0.921951                  0.693117               0.700552   
3                 0.408469                  0.764419               0.641189   
4                 0.615968                  0.623071               0.560186   
5                 0.859174                  0.375240               0.392039   
6                 0.736465                  0.408850               0.399057   
7                 1.000000                  0.548314               0.580071   
8                 0.383189                  0.951942               0.867237   
9                 0.061407                  0.839805               0.720730   
10                0.000000                  0.661392               0.971927   
11                0.890984                  0.532294               0.755236   
12                0.936170                  0.606423               0.959352   
13                0.886876                  0.549570               0.938297   
14                0.954708                  0.542031               0.771612   
15                0.8

Berdasarkan output tersebut, dapat dilihat bahwa setiap provinsi berhasil
dikelompokkan ke dalam 4 cluster yang berbeda, yaitu cluster 1, 2, 3 dan 4. Hal ini menunjukkan bahwa algoritma K-Means berhasil mengelompokkan provinsi-provinsi berdasarkan kesamaan karakteristiknya.

### **F.1.3 Silhouette Plot**

In [ ]:
import numpy as np
import pandas as pd
import plotly.graph_objects as go
from sklearn.metrics import silhouette_samples, silhouette_score
from sklearn.cluster import KMeans

# Hitung silhouette score dan nilai untuk setiap titik
silhouette_avg = silhouette_score(X, cluster_labels)
silhouette_values = silhouette_samples(X, cluster_labels)

# Siapkan data untuk visualisasi
n_clusters = len(np.unique(cluster_labels))
silhouette_data = pd.DataFrame({
    'Cluster': cluster_labels + 1,
    'Silhouette Coefficient': silhouette_values
})

# Urutkan data berdasarkan klaster dan nilai silhouette
silhouette_data = silhouette_data.sort_values(by=['Cluster', 'Silhouette Coefficient'], ascending=[True, True])

# Tambahkan kolom posisi untuk membantu visualisasi
silhouette_data['Posisi'] = np.arange(len(silhouette_data))

# Membuat plot menggunakan Plotly
fig = go.Figure()

# Loop untuk setiap klaster
for cluster_id in range(1, n_clusters + 1):
    cluster_data = silhouette_data[silhouette_data['Cluster'] == cluster_id]
    fig.add_trace(go.Bar(
        x=cluster_data['Silhouette Coefficient'],
        y=cluster_data['Posisi'],
        orientation='h',
        marker=dict(color=px.colors.qualitative.Plotly[cluster_id % len(px.colors.qualitative.Plotly)]),
        name=f'Cluster {cluster_id}'
    ))

# Garis rata-rata silhouette score
fig.add_vline(
    x=silhouette_avg,
    line=dict(color='red', dash='dash'),
    annotation_text="Rata-rata Silhouette",
    annotation_position="top right"
)

# Layout
fig.update_layout(
    width=1200,
    height=600,
    title=dict(
        text="<b>Silhouette Plot</b><br><sup><sup>dengan Algoritma K-Means</sup></sup>",
        font=dict(size=25, color='#757882'),
    ),
    xaxis=dict(
        title="Nilai Silhouette Coefficient",
        linecolor="black",
        showgrid=False,
        zeroline=True
    ),
    yaxis=dict(
        title="Cluster",
        showticklabels=False,
        linecolor="black",
        showgrid=False
    ),
    plot_bgcolor='rgba(0, 0, 0, 0)',
    showlegend=False,
    margin=dict(l=50)
)

fig.show(renderer='colab')


Dari hasil Silhouette Plot, rata-rata Silhouette Score bernilai **0.37** nilai ini menunjukkan pengelompokkan cukup baik tetapi tidak optimal, karena mendekati batas bawah moderat.

### **F.1.4 Plot Hasil Cluster**

In [ ]:
import plotly.graph_objects as go

# Membuat plot 3D
fig = go.Figure(
    data = [
        go.Scatter3d(
            x = data_cluster_2022['ketersediaan'],
            y = data_cluster_2022['keterjangkauan'],
            z = data_cluster_2022['pemanfaatan'],
            mode = 'markers',
            marker = dict(
                size = 5,
                color = data_cluster_2022['cluster_kmeans'],
                opacity = 0.5,
                colorscale='Viridis',
                colorbar = dict(title='Color Legend')
            ),
            hovertext = data_cluster_2022['provinsi'],
            hovertemplate='<b>%{hovertext}</b><br>ketersediaan: %{x}<br>Keterjangkauan: %{y}<br>Pemanfaatan: %{z}<br>Cluster: %{marker.color}'
        )
    ]
)


fig.update_layout(
    width = 1200,
    height = 800,
    plot_bgcolor = 'rgba(0, 0, 0, 0)',
    title = dict(
        text = '',
        font = dict(
            size = 20
        )
    ),
    scene = dict(
        xaxis_title = 'Ketersediaan',
        yaxis_title = 'Keterjangkauaan',
        zaxis_title = 'Pemanfaatan',
        xaxis=dict(showgrid=False),
        yaxis=dict(showgrid=False),
        zaxis=dict(showgrid=False)
    )
)

# Tampilkan plot
fig.show()

Proses clustering ini menghasilkan empat kelompok (cluster) berbeda berdasarkan karakteristik kesamaan dalam ketiga aspek tersebut. Setelah dilakukan clustering, setiap provinsi diberi label sesuai dengan kelompoknya. Berdasarkan hasil clustering, berikut adalah pembagian provinsi dalam masing-masing cluster:

Cluster 1 (Ungu): Daerah dengan ketersediaan dan keterjangkauan yang tinggi serta pemanfaatan yang stabil, meskipun terdapat variasi sedang pada ketersediaan.

Cluster 2 (Biru): Daerah dengan keterjangkauan yang sangat tinggi dan konsisten, tingkat pemanfaatan yang baik, namun tingkat ketersediaan masih rendah.

Cluster 3 (Hijau): Daerah dengan ketersediaan, keterjangkauan, dan pemanfaatan menunjukkan tingkat yang tinggi pada ketiganya, dengan variasi kecil hingga sedang.

Cluster 4 (Kuning) : Daerah dengan tingkat ketersediaan yang sangat rendah, keterjangkauan yang moderat tanpa variasi, dan pemanfaatan yang rendah hingga sedang dengan variasi kecil.

Hasil ini dapat digunakan untuk mengidentifikasi daerah yang membutuhkan perhatian khusus seperti daerah di Cluster 4.
Selain itu, cluster dengan nilai tinggi seperti Cluster 3 dapat dijadikan contoh atau perbandingan untuk daerah lainnya.

### **F.1.5 Summary**

In [ ]:
# Menghitung mean, median, dan simpangan baku untuk setiap cluster
cluster_statistics = data_cluster_2022.groupby('cluster_kmeans').agg({
    'ketersediaan': ['mean', 'median', 'std'],
    'keterjangkauan': ['mean', 'median', 'std'],
    'pemanfaatan': ['mean', 'median', 'std']
}).reset_index()

# Menampilkan hasil
display(cluster_statistics)

cluster_kmeans ketersediaan                    keterjangkauan            \
                         mean  median        std           mean    median   
0              1    73.995000  80.180  14.776302      70.730625  71.72500   
1              2    30.148750  37.580  20.416207      88.817500  89.44500   
2              3    83.006667  84.395   9.921597      82.418333  81.44500   
3              4    14.745000  14.745   4.150717      62.163750  62.16375   

            pemanfaatan                       
        std        mean     median       std  
0  5.091352   62.721667  61.900000  2.181361  
1  2.905540   72.600000  72.370000  6.770172  
2  4.405731   71.025000  71.320000  6.342421  
3  0.000000   52.026875  52.026875  5.859264

**Cluster 1 (Ungu)**

Ketersediaan rata-rata 73,99 (median 80,18), keterjangkauan 70,73 (median 71,73), dan pemanfaatan 62,72 (median 61,90). Menunjukkan wilayah dengan ketersediaan tinggi, keterjangkauan baik, dan pemanfaatan stabil, mencerminkan infrastruktur atau akses yang baik.


**Cluster 2 (Biru)**

Ketersediaan rata-rata 30,15 (median 37,58), keterjangkauan 88,82 (median 89,45), dan pemanfaatan 72,60 (median 72,37). Menunujukkan wilayah dengan keterjangkauan sangat tinggi, namun ketersediaan rendah membatasi, meskipun pemanfaatan cukup baik berkat optimasi sumber daya.

**Cluster 3 (Hijau)**

Ketersediaan rata-rata 83,01 (median 84,39), keterjangkauan 82,42 (median 81,45), dan pemanfaatan 71,03 (median 71,32). Menunujukkan wilayah dengan keseimbangan optimal antara ketersediaan, keterjangkauan, dan pemanfaatan sumber daya.

**Cluster 4 (Kuning)**

Ketersediaan rata-rata 14,75, keterjangkauan 62,16, dan pemanfaatan 52,03 (dengan median yang sama)  Hal ini mengindikasikan area ini memiliki akses yang terbatas, meskipun keterjangkauannya stabil.